File: days/day_09/main.py
import time
import os

def part_one(nums):
	i = 25
	
	while i < len(nums):
		current = nums[i]
		found = False
		S = nums[i-25:i]

		for n in S:
			for m in S:
				if n+m == current and n != m:
					found = True
		
		if not found:
			return nums[i]

		i+=1


def part_two(nums, N):
	for i in range(len(nums)):
		k = 0
		
		while sum(S := nums[i:i+k])<N:
			k+=1
		
		if sum(S) == N:
			return min(S) + max(S)


def main():
	start_time = time.time()

	with open(os.path.dirname(__file__) + '/input.txt', 'r') as data:
		nums = [int(line) for line in data.readlines()]
		
		part_one_ans = part_one(nums)
		part_two_ans = part_two(nums, part_one_ans)

		print('day  9 ({:,.3f}s)'.format(time.time()-start_time))
		print('  part 1: {}'.format(part_one_ans))
		print('  part 2: {}'.format(part_two_ans))

if __name__ == "__main__":
		main()

File: days/day_09/main.ml
open Core
open Core_kernel

let file = "input.txt"

let part_one nums =
	List.fold_until nums
		~init:[] 
		~f:(fun prev x -> 
			if (List.length prev) < 25 
			then Continue (prev @ [x])	
			else 
				if List.for_all prev ~f:(fun n ->
						match 
							List.find prev ~f:(fun m -> n + m = x && not (n = m)) 
						with
						| Some _m -> false
						| None -> true )
				then Stop x
				else Continue ((List.tl_exn prev) @ [x]))
		~finish:(fun _prev -> 0)
;;

let part_two nums n =
	let rec search remainder summands total =
	if total < n then
		match remainder with
		| hd :: tl -> 
			search 
				tl 
				(Fdeque.enqueue summands `back hd) 
				(total + hd)
		| [] -> []
	else if total > n then 
		match (Fdeque.dequeue summands `front) with
		| Some (hd, tl) -> 
			search
				remainder 
				tl
				(total - hd)
		| None -> []
	else
		Fdeque.to_list summands
	in
	let summands =  (search nums Fdeque.empty 0) in
	let compare = (-) in
	match (
		List.max_elt summands ~compare, 
		List.min_elt summands ~compare) 
	with
		| (Some max, Some min) -> max + min
		| _ -> 0
;;

let () = 
	let start_time = Core.Unix.gettimeofday () in
	let nums = file
		|> In_channel.read_lines
		|> List.map ~f:int_of_string
	in
	let n = (part_one nums) in
	printf "day  9 (%.3fs)\n" ((Core.Unix.gettimeofday ()) -. start_time);
	printf "  part 1: %d\n" n;
	printf "  part 2: %d\n" (part_two nums n)
;;
